src/core/autotracker.h

```diff
@@ -288,18 +348,65 @@ class AutoTracker final : public LuaInterface<AutoTracker>{
             for (size_t i=0; i<len; i++)
                 res.push_back(buf[i]);
         }
+        if( _provider ) {
+            uint8_t buf[len];
+            _provider->readFromCache((uint32_t)addr, len, buf);
+            for( size_t i = 0; i < len; i++ )
+                res.push_back(buf[i]);
+        }
         return res;
     }
     
     int ReadU8(int segment, int offset=0)
     {
+      if( _provider )
+      {
+        // this is a live blocking call to read memory from the game
+        uint32_t address = segment;
+        uint32_t o = offset;
+        return _provider->readU8Live(address, o);
```

Comment:
In order to fully eliminate live reads, wouldn't we have to guarantee that the entire cache is always updated in an atomic manner? I rely on `AutoTracker:ReadU8` and `AutoTracker:ReadU16` in Hamsda's pack for rare cases where I need to read memory from outside of my segment and there is no guarantee that what's cached there is going to match what I have in my segment. If we remove the option for live reads here, the behavior is not going to match Emotracker and the pack will not work as expected.

Reply:
Basically, a pack should not expect the memory to be consistent with the protocols we have. This is less of a problem with emulators, and less of a problem on localhost, but basically a NMI or a soft reset (or even a hard reset) can happen at any time between two resets.

Anyway, I'm open to a stop gap solution by implementing some stuff "as close as possible", but in case of blocking reads, they need a short timeout, so the tracker does not freeze when opening the menu. The backlog of replies that happened async will have to be dropped, or the connector will have to drop the connection to properly report a "reset" of the socket.

Reply:
Quick question, are you doing anything with the return value from the memory watch callback functions? In Emotracker, the segment callback will be called once for every change to the cache, unless the callback returned false last time in which case it will be called again on the next loop.

Reply:
I changed everything over to async tcp, no calls are made on the main thread anymore. There are a couple of consequences for this, most of which I should be able to handle by fixing the pack.

Please see [c95ed85](https://github.com/black-sliver/PopTracker/pull/52/commits/c95ed85be9379743f6dbfa56c26ef47fbb9a8a12), I made a change to ScriptHost to run all callbacks when any data is changed, instead of only the one. This allows callbacks to be re-executed when necessary. The dirty flag can be cleared when the callback does not return false, which prevents it from running again needlessly. I left a TODO in there to check the return value, which I don't know how to implement.

Reply:
I'll check the whole code a bit later. I should have time this weekend :-)

But Re: edit: As far as I know, it's not safe to enter the same Lua_State from two threads at the same time. Besides that, calling from Lua into native code would also require locking (you can't access the UI from two threads at the same time). What would work is using a GIL and having 3 threads: UI, Network, Dispatch where UI runs the "front-end" Lua, Network only pushes to a queue and Dispatch runs the "aync" Lua; UI and Dispatch lock the Lua_State so only one can enter it at the same time, but that's hardly any better than having the UI thread run the dispatch code. Also care has to be taken when stepping into Lua from access rules and the like - especially the load_* functions can invalidate a lot of stuff (Lists of widgets, for example).

To digress a bit more, I have not looked into details, but I think queueing a task in Lua's cooperative multi-tasking/threading does not solve it either, because that only gets run while Lua is executing - so to make *that* work, would mean Lua would spin in its own thread and we'd need to design the rest around this, which includes locking in a lot of places.